File: javascript/leetcode/increasing-order-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
const increasingBST = function(root) {
    const values = traverse(root, []);
    return buildNewTree(values);
};

const traverse = function(root, values) {
    if (root === null) { return values }
    values = traverse(root.left, values);
    values.push(root.val);
    values = traverse(root.right, values);
    return values;
}

const buildNewTree = function(values) {
    const newRoot = new TreeNode();
    let root = newRoot;
    let i = 0;
    while (i < values.length) {
        root.val = values[i];
        if (i < values.length - 1) {
            root.right = new TreeNode();
            root = root.right;
        }
        i++;
    }
    return newRoot;
}


/** Alternative relinking solution
 * @param {TreeNode} root
 * @return {TreeNode}
 */
const increasingBST = function(root) {
    const head = new TreeNode();
    let curr = head;
    const traverse = function(root) {
        if (root === null) { return }
        traverse(root.left);
        curr.right = root;
        curr = root
        curr.left = null;
        traverse(root.right);
    }
    traverse(root);
    return head.right;
};

File: python/leetcode/increasing-order-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    head = TreeNode(None)
    curr = head
    def increasingBST(self, root: TreeNode) -> TreeNode:
        self.traverse(root)
        return self.head.right

    def traverse(self, root: TreeNode) -> None:
        if root is None:
            return
        self.traverse(root.left)
        self.curr.right = root
        self.curr = root
        self.curr.left = None
        self.traverse(root.right)
